include/swift/Remote/MetadataReader.h

```diff
@@ -812,22 +813,17 @@ class MetadataReader {
         sizeof(TargetFunctionTypeMetadata<Runtime>);
       for (StoredPointer i = 0; i < Function->getNumArguments(); ++i,
            ArgumentAddress += sizeof(StoredPointer)) {
-        StoredPointer FlaggedArgumentAddress;
+        StoredPointer ParameterAddress;
         if (!Reader->readInteger(RemoteAddress(ArgumentAddress),
-                                 &FlaggedArgumentAddress))
+                                 &ParameterAddress))
           return BuiltType();
 
         FunctionParam<BuiltType> Param;
+        if (auto ParamTypeRef = readTypeFromMetadata(ParameterAddress)) {
+          auto ParameterFlags = Function->getParameterFlags(i);
 
-        // TODO: Use target-agnostic FlaggedPointer to mask this!
-        const auto InOutMask = (StoredPointer) 1;
-        // FIXME: Add import parameter related flags from metadata
-        if ((FlaggedArgumentAddress & InOutMask) != 0)
-          Param.setInOut();
-
-        FlaggedArgumentAddress &= ~InOutMask;
-        if (auto ParamTypeRef = readTypeFromMetadata(FlaggedArgumentAddress)) {
           Param.setType(ParamTypeRef);
+          Param.setFlags(ParameterTypeFlags::fromRaw(ParameterFlags));
```

Comment:
It is really not okay to assume that the AST's ParameterTypeFlags type matches the runtime-metadata type.  Please make these separate types with their own internal representation of all the various flags.  I'm sure I asked you to do this before.

Reply:
Ok, will do.